### ISSUE TITLE
'thrust::system::system_error' what(): parallel_for failed: out of memory Aborted

### ISSUE THREAD

**Issue by appleyuchi** [dmlc/xgboost#4898]
I feed  datasets(1.09 million cases contained)  to xgboost with GPU support,
and get the error：
'thrust::system::system_error' what(): parallel_for failed: out of memory Aborted

Which means that the GPU-Memory is OUT of use.

So is there any method to feed data to xgboost piece after piece?

**Comment by hcho3** [comment]
@sriramch and @rongou are working on external memory support with GPUs.

**Comment by sriramch** [comment]
@appleyuchi - please enable external memory option and try. [here](https://github.com/dmlc/xgboost/blob/master/doc/tutorials/external_memory.rst) is a start.

how many features does your dataset have? how big is your training file? how many gpus, it size and how much system ram do you have?

**Comment by appleyuchi** [comment]
@sriramch 
In Kaggl IEEE-CIS-Fraud-Detection.
1.09millon datasets totally.
When Use Semi-Supervision Training.
AND....
The GPU is on Nvidia-V100 -16G

**Comment by appleyuchi** [comment]
what's the meaning of gpus？

**Comment by appleyuchi** [comment]
The link you provided seems hard,I can not follow it.

**Comment by sriramch** [comment]
@appleyuchi what part of the document do you find it hard to follow? it should be possible to handle the dataset of that size, given the gpu you have. how much of system ram do you have? if you need to enable training with external memory, then you simply append the '#' symbol to the training file and provide the cache file location. the dataset of this size shouldn't require external memory enablement.

also provide the training configuration along with how you are invoking the xgboost api's.

btw, i'm looking at [this](https://www.kaggle.com/c/ieee-fraud-detection/data) for the dataset.

**Comment by trivialfis** [comment]
It could happen in extremely wide dataset.

> what's the meaning of gpus？

More than 1 GPUs I think.

> The link you provided seems hard,I can not follow it.

Which part you found difficult?  I can help improving the doc.

**Comment by appleyuchi** [comment]
Sorry for late reply.@sriamch @trivialfis 
I mean:
https://github.com/dmlc/xgboost/blob/master/doc/tutorials/external_memory.rst
this linke provided by @sriramch 

Why so much strange words in this link?
Why `libsvm` is mentioned in this link ?
What's the meaning of "filename#cacheprefix"?
I just use .csv file ,what's the relationship with my dataset(.csv file) and "filename#cacheprefix"

I could not follow this document,is there a totally integral example?
Much Thanks.

**Comment by trivialfis** [comment]
@appleyuchi 

`libsvm` means the dataset stored in `libsvm` file format (the file format used by libsvm).

**Comment by trivialfis** [comment]
For the cache prefix, for example, if you have a dataset stored in file `dataset.txt` with libsvm format.  You can use
``` python
dtrain = DMatrix('./dataset.txt#dtrain.cache')
```
XGBoost will load `dataset.txt` in, preprocess it, then write to a new file named `dtrain.cache` as an on disk cache for storing preprocessed data in binary form.

**Comment by trivialfis** [comment]
The strings and weird symbols are specified by uri format https://en.wikipedia.org/wiki/Uniform_Resource_Identifier .  It appears on top of every browsers when you browse the internet.  Will submit a PR for more detail later.

**Comment by appleyuchi** [comment]
Got it,Much Thanks @trivialfis
